src/meta/meta_turtle.c

```diff
@@ -81,12 +94,76 @@ __metadata_load_hot_backup(WT_SESSION_IMPL *session)
         WT_ERR(__wt_getline(session, fs, value));
         if (value->size == 0)
             WT_ERR_PANIC(session, EINVAL, "%s: zero-length value", WT_METADATA_BACKUP);
+        /*
+         * When performing partial backup restore, check through the file metadata entries and check
+         * if the file exists in the file system. If the file doesn't exist, append the filename to
+         * the partial backup list and the partial backup remove list so we can clean up the
+         * metadata and history store afterwards.
+         */
+        filename = (char *)key->data;
+        if (restore_partial_backup && WT_PREFIX_SKIP(filename, "file:")) {
+            WT_ERR(__wt_fs_exist(session, filename, &exist));
+            if (!exist) {
+                WT_ERR(__wt_realloc_def(
+                  session, &allocated_int, slot + 1, &conn->partial_backup_remove_list));
+                /* Leave a NULL at the end to mark the end of the list. */
+                WT_ERR(__wt_realloc_def(session, &allocated, slot + 2, &partial_backup_list));
+                p = &partial_backup_list[slot];
+                p[0] = p[1] = NULL;
+
+                WT_ERR(__wt_strdup(session, (char *)key->data, p));
```

Comment:
I'm not sure if we're guaranteed that key->data is null terminated. You may need to use strndup with the length.